packages/assets-controllers/src/MultichainBalancesController/MultichainBalancesController.ts

```diff
@@ -0,0 +1,390 @@
+    );
+    this.messagingSystem.subscribe(
+      'AccountsController:accountRemoved',
+      (account) => this.#handleOnAccountRemoved(account),
+    );
+  }
+
+  /**
+   * Starts the polling process.
+   */
+  start(): void {
+    this.#tracker.start();
+  }
+
+  /**
+   * Stops the polling process.
+   */
+  stop(): void {
+    this.#tracker.stop();
+  }
+
+  /**
+   * Gets the block time for a given account.
+   *
+   * @param account - The account to get the block time for.
+   * @returns The block time for the account.
+   */
+  #getBlockTimeFor(account: InternalAccount): number {
+    if (account.type in BALANCE_UPDATE_INTERVALS) {
+      return BALANCE_UPDATE_INTERVALS[
+        account.type as keyof typeof BALANCE_UPDATE_INTERVALS
+      ];
+    }
+    throw new Error(
+      `Unsupported account type for balance tracking: ${account.type}`,
+    );
+  }
+
+  /**
+   * Lists the multichain accounts coming from the `AccountsController`.
+   *
+   * @returns A list of multichain accounts.
+   */
+  #listMultichainAccounts(): InternalAccount[] {
+    return this.messagingSystem.call(
+      'AccountsController:listMultichainAccounts',
+    );
+  }
+
+  /**
+   * Lists the accounts that we should get balances for.
+   *
+   * @returns A list of accounts that we should get balances for.
+   */
+  #listAccounts(): InternalAccount[] {
+    const accounts = this.#listMultichainAccounts();
+
+    return accounts.filter((account) => this.#isNonEvmAccount(account));
+  }
+
+  /**
+   * Get a non-EVM account from its ID.
+   *
+   * @param accountId - The account ID.
+   * @returns The non-EVM account.
+   */
+  #getAccount(accountId: string): InternalAccount {
+    const account: InternalAccount | undefined =
+      this.#listMultichainAccounts().find(
+        (multichainAccount) => multichainAccount.id === accountId,
+      );
+
+    if (!account) {
+      throw new Error(`Unknown account: ${accountId}`);
+    }
+    if (!this.#isNonEvmAccount(account)) {
+      throw new Error(`Account is not a non-EVM account: ${accountId}`);
+    }
+    return account;
+  }
+
+  /**
+   * Updates the balances of one account. This method doesn't return
+   * anything, but it updates the state of the controller.
+   *
+   * @param accountId - The account ID.
+   */
+  async #updateBalance(accountId: string) {
+    const account = this.#getAccount(accountId);
+    const partialState: MultichainBalancesControllerState = { balances: {} };
+
+    if (account.metadata.snap) {
+      const scope = getScopeForAccount(account);
+      const assetTypes = NETWORK_ASSETS_MAP[scope];
+
+      partialState.balances[account.id] = await this.#getBalances(
+        account.id,
+        account.metadata.snap.id,
+        assetTypes,
+      );
+    }
+
+    this.update((state: Draft<MultichainBalancesControllerState>) => {
+      state.balances = {
+        ...state.balances,
+        ...partialState.balances,
+      };
+    });
+  }
```

Comment:
```suggestion
  async #updateBalance(accountId: string) {
    const account = this.#getAccount(accountId);

    if (account.metadata.snap) {
      const scope = getScopeForAccount(account);
      const assetTypes = NETWORK_ASSETS_MAP[scope];

      const accountBalance = await this.#getBalances(
        account.id,
        account.metadata.snap.id,
        assetTypes,
      );

      this.update((state: Draft<MultichainBalancesControllerState>) => {
        state.balances[accountId] = accountBalance;
      });
    }
  }
```

Reply:
done in [bcaee6c](https://github.com/MetaMask/core/pull/4965/commits/bcaee6c42b4c196809fe2c59cd611ca155794b28)